### ISSUE TITLE
Better support for bigints in generic functions

### ISSUE THREAD

**Issue by vks** [rust-num/num#273]
Unlike integer types that are `Copy`, for bigints it is important to avoid cloning by supporting `x + &y`, `&x + &y` etc. for all relevant operations and comparisons. The required trait bounds are cumbersome, because they need the special `for<'a>` syntax. The best way to deal with that is in my opinion what @andersk proposed:

```rust
fn extended_gcd<T: Integer>(a: T, b: T) -> GcdResult<T>
    where for<'a> &'a T: IntegerOps<T>
```

for some appropriately defined trait [`IntegerOps`](https://github.com/vks/discrete-log/blob/master/src/main.rs#L104).

This would be useful for optimizations like in #153, i.e. for generic functions like `pow`, `gcd` and `binomial`.

**Comment by vks** [comment]
I believe this was implemented in 21b520ea15a07b2197c4c6950a4996a182b56619. The functions above don't use it yet, but this would be a breaking change.

**Comment by cuviper** [comment]
This issue was moved to rust-num/num-bigint#2
